core/common/src/main/java/alluxio/security/LoginUser.java

```diff
@@ -50,18 +50,21 @@ private LoginUser() {} // prevent instantiation
    * @return the login user
    * @throws IOException if login fails
    */
-  public static User get() throws IOException {
+  public static synchronized User get() throws IOException {
```

Comment:
Re: volatile, the article has a section saying:
> A commonly suggested nonfix is to declare the resource field of SomeClass as volatile. However, while the JMM prevents writes to volatile variables from being reordered with respect to one another and ensures that they are flushed to main memory immediately, it still permits reads and writes of volatile variables to be reordered with respect to nonvolatile reads and writes. That means -- unless all Resource fields are volatile as well -- thread B can still perceive the constructor's effect as happening after resource is set to reference the newly created Resource.

I don't see how to do this with AtomicReference - if two threads see that login hasn't happened yet, they will both try to log in at the same time. Moreover, the overhead of using AtomicReference is probably similar to that of synchronized.

The right way to do this is to create the logged-in User at the start and pass the `User` around to the places that need it, but this would be a significant change requiring a day or two of work. Alternately we could either create the `User` eagerly when the class loads, or ignore the race condition since it's very unlikely to actually come up. I prefer ignoring between these two options since doing expensive operations during class loading is more likely to lead to its own class of painful bugs.